File: backend/src/avito_reviewer/ai/rubric.py
"""Рубрика как данные.

Промпт генерируется из этой структуры, а не пишется руками под каждое
задание: добавить курс значит добавить JSON. Схема должна вмещать все формы,
в которых условия приходят на самом деле, — таблицу с баллами, N вопросов по
фиксированному баллу, чек-лист без баллов вовсе, требования к формату со
штрафами.

Три поля появились из разбора реальных условий и объясняют, почему нельзя
свести оценку к сумме:

* `min_score_for_pass` — в системном дизайне провал по обязательному критерию
  не компенсируется набранным на остальных;
* `step` — баллы дробные, и округлять надо по правилу задания;
* `checks[]` — «проверочные пункты для преподавателя» из условия; модель
  отвечает по каждому отдельно, и разброс между прогонами резко падает.

Заполняет рубрику методист (в перспективе — Rubric Compiler с подтверждением
человеком). Здесь только схема, разбор и реестр.
"""

from __future__ import annotations

import json
import logging
import re
from datetime import UTC, datetime, timedelta
from pathlib import Path
from typing import Any, Literal

from pydantic import AliasChoices, BaseModel, Field, field_validator

log = logging.getLogger(__name__)


class Scale(BaseModel):
    """Шкала задания: максимум, порог зачёта и шаг округления."""

    total_max: float
    pass_threshold: float | None = None
    step: float = 1.0

    def round_to_step(self, value: float) -> float:
        if self.step <= 0:
            return value
        return round(round(value / self.step) * self.step, 4)


class LatePolicy(BaseModel):
    """Штраф за просрочку — правило из условия, а не наша выдумка.

    «Досдача в течение 1 дня после дедлайна — −1 балл; позже — 0 баллов»
    ложится сюда как `grace_days=1, penalty_per_grace_day=1, after_grace=zero`.
    """

    grace_days: int = 0
    penalty_per_grace_day: float = Field(
        default=0.0,
        validation_alias=AliasChoices("penalty_per_grace_day", "penalty_per_grace"),
    )
    after_grace: Literal["zero", "continue"] = Field(
        default="zero", validation_alias=AliasChoices("after_grace", "after")
    )

    def apply(
        self, score: float, submitted_at: datetime | None, deadline_at: datetime | None
    ) -> tuple[float, str]:
        if not submitted_at or not deadline_at:
            return score, "сдано в срок"

        submitted, deadline = _aligned(submitted_at, deadline_at)
        if submitted <= deadline:
            return score, "сдано в срок"

        days = max(1, -(-(submitted - deadline) // timedelta(days=1)))
        if days <= self.grace_days:
            penalty = self.penalty_per_grace_day * days
            return max(0.0, score - penalty), f"просрочка {days} дн., штраф −{penalty:g}"
        if self.after_grace == "zero":
            return 0.0, f"просрочка {days} дн. — 0 баллов"
        penalty = self.penalty_per_grace_day * days
        return max(0.0, score - penalty), f"просрочка {days} дн., штраф −{penalty:g}"


def _aligned(left: datetime, right: datetime) -> tuple[datetime, datetime]:
    """Привести срок и время сдачи к сравнимому виду.

    `submitted_at` приходит от провайдера и всегда со смещением, а `deadline_at`
    — из тела запроса, где клиент может прислать «2026-02-09T23:59» без зоны.
    Сравнение таких дат роняет запрос целиком, поэтому наивную дату читаем как
    UTC: ошибиться на смещение лучше, чем отдать 500 на штатном вводе.
    """
    if (left.tzinfo is None) == (right.tzinfo is None):
        return left, right
    return (
        left if left.tzinfo else left.replace(tzinfo=UTC),
        right if right.tzinfo else right.replace(tzinfo=UTC),
    )


class FormatCheck(BaseModel):
    """Формальное требование из условия — проверяется кодом, без единого токена.

    Format Gate ещё не построен; рубрики уже несут его правила, потому что
    вынимать их из условия надо один раз вместе с критериями, а не потом.
    """

    check: str
    level: Literal["blocking", "warning", "info"] = "warning"
    """`blocking` — к ревьюеру без прогона модели; `warning` — флаг в карточке;
    `info` — факт в контекст ревью-агента, чтобы он не пересчитывал его сам."""
    params: dict[str, Any] = Field(default_factory=dict)
    note: str = ""


class Criterion(BaseModel):
    id: str
    title: str
    max_score: float
    min_score_for_pass: float | None = None
    weight: float = 1.0
    description: str = ""
    checks: list[str] = Field(default_factory=list)
    anchors: dict[str, str] = Field(default_factory=dict)
    evidence_required: bool = True
    auto_verifiable: bool = False
    """Проверяется детерминированно (Format Gate, smoke-тесты), а не моделью."""
    ai_sensitive: bool = False
    """Критерий про самостоятельность: сигнал детектора здесь показывается первым."""

    @field_validator("anchors", mode="before")
    @classmethod
    def _stringify_keys(cls, value: Any) -> Any:
        if isinstance(value, dict):
            return {str(key): str(text) for key, text in value.items()}
        return value


class Rubric(BaseModel):
    assignment_id: str
    title: str = ""
    course: str = ""
    stage: str | None = None
    source_note: str = ""
    ai_policy: str = "declare_required"

    scale: Scale
    late_policy: LatePolicy = Field(default_factory=LatePolicy)
    format_gate: list[FormatCheck] = Field(default_factory=list)
    criteria: list[Criterion] = Field(default_factory=list)

    @field_validator("format_gate", mode="before")
    @classmethod
    def _normalize_gate(cls, value: Any) -> Any:
        """Принять и плоский список, и группировку по уровню из архитектуры."""
        if isinstance(value, dict):
            return [
                {**check, "level": level}
                for level in ("blocking", "warning", "info")
                for check in value.get(level, [])
                if isinstance(check, dict)
            ]
        return value

    def criterion(self, criterion_id: str) -> Criterion | None:
        return next((c for c in self.criteria if c.id == criterion_id), None)

    @property
    def ai_sensitive_criteria(self) -> list[Criterion]:
        return [c for c in self.criteria if c.ai_sensitive]


ID_PATTERN = re.compile(r"^[a-zA-Z0-9][a-zA-Z0-9._-]{0,63}$")


def validate_rubric(rubric: Rubric) -> list[str]:
    """Что не так с рубрикой. Пустой список — можно принимать.

    Проверяет код, а не модель, и проверяет то, что ломает подсчёт балла на
    каждой работе потока: рубрика действует до конца курса, и арифметическая
    ошибка в ней стоит дороже любой ошибки в одном черновике.
    """
    problems: list[str] = []

    if not ID_PATTERN.match(rubric.assignment_id):
        problems.append(
            f"идентификатор {rubric.assignment_id!r} не годится для имени файла: "
            f"латиница, цифры, точка, дефис и подчёркивание"
        )
    if not rubric.criteria:
        problems.append("в рубрике нет ни одного критерия")

    ids = [c.id for c in rubric.criteria]
    duplicates = sorted({i for i in ids if ids.count(i) > 1})
    if duplicates:
        problems.append("повторяющиеся идентификаторы критериев: " + ", ".join(duplicates))

    if rubric.scale.total_max <= 0:
        problems.append("максимальный балл должен быть больше нуля")
    if rubric.scale.step < 0:
        problems.append("шаг шкалы не может быть отрицательным")

    threshold = rubric.scale.pass_threshold
    if threshold is not None and threshold > rubric.scale.total_max:
        problems.append(
            f"порог зачёта {threshold:g} выше максимума {rubric.scale.total_max:g} — "
            f"зачёт недостижим"
        )

    for criterion in rubric.criteria:
        if criterion.max_score <= 0:
            problems.append(f"{criterion.id}: максимум критерия должен быть больше нуля")
        minimum = criterion.min_score_for_pass
        if minimum is not None and minimum > criterion.max_score:
            problems.append(
                f"{criterion.id}: обязательный минимум {minimum:g} выше максимума "
                f"{criterion.max_score:g} — критерий провален всегда"
            )

    # Сумма может не совпадать с максимумом: у критериев бывают веса. А вот
    # недостижимый максимум — это уже поломка шкалы.
    reachable = round(sum(c.max_score * (c.weight or 1.0) for c in rubric.criteria), 4)
    if rubric.criteria and reachable < rubric.scale.total_max:
        problems.append(
            f"максимум {rubric.scale.total_max:g} недостижим: по всем критериям "
            f"с весами набирается {reachable:g}"
        )
    return problems


def load_rubric(path: str | Path) -> Rubric:
    return Rubric.model_validate(json.loads(Path(path).read_text(encoding="utf-8")))


class RubricRejected(ValueError):
    """Рубрика не прошла проверку и в каталог не попала."""

    def __init__(self, problems: list[str]) -> None:
        super().__init__("; ".join(problems))
        self.problems = problems


class RubricExists(ValueError):
    """Рубрика с таким идентификатором уже есть."""

    def __init__(self, assignment_id: str) -> None:
        super().__init__(f"рубрика {assignment_id!r} уже существует")
        self.assignment_id = assignment_id


class RubricStore:
    """Рубрики из каталога JSON-файлов.

    Пока рубрики не переехали в базу и админку, каталог — и есть источник
    правды: `rubrics/<assignment>.json`, ключ — `assignment_id`.
    """

    def __init__(self, directory: str | Path) -> None:
        self.directory = Path(directory)
        self._rubrics: dict[str, Rubric] = {}
        self.reload()

    def reload(self) -> None:
        self._rubrics = {}
        if not self.directory.is_dir():
            log.warning("rubrics: %s is not a directory — no rubrics loaded", self.directory)
            return
        for path in sorted(self.directory.glob("*.json")):
            try:
                rubric = load_rubric(path)
            except (ValueError, TypeError, OSError) as exc:
                log.warning("rubrics: skipping %s — %s", path.name, exc)
                continue
            self._rubrics[rubric.assignment_id] = rubric
        log.info("rubrics: loaded %d from %s", len(self._rubrics), self.directory)

    def get(self, assignment_id: str) -> Rubric | None:
        return self._rubrics.get(assignment_id)

    def save(self, rubric: Rubric, *, overwrite: bool = False) -> Path:
        """Положить подтверждённую рубрику в каталог.

        Молча переписать существующую нельзя: по ней уже могли быть проверены
        работы, и подмена задним числом делает их баллы необъяснимыми.
        """
        problems = validate_rubric(rubric)
        if problems:
            raise RubricRejected(problems)

        path = self.directory / f"{rubric.assignment_id}.json"
        if path.exists() and not overwrite:
            raise RubricExists(rubric.assignment_id)

        self.directory.mkdir(parents=True, exist_ok=True)
        path.write_text(
            rubric.model_dump_json(indent=2, exclude_defaults=False), encoding="utf-8"
        )
        self._rubrics[rubric.assignment_id] = rubric
        log.info("rubrics: saved %s to %s", rubric.assignment_id, path)
        return path

    def delete(self, assignment_id: str) -> None:
        """Remove a rubric from the catalogue. Submissions already scored against
        it keep meaning what they meant — `Submission.rubric_snapshot` froze the
        rubric at scoring time, so deleting the file here does not touch them.
        """
        if assignment_id not in self._rubrics:
            raise KeyError(assignment_id)
        path = self.directory / f"{assignment_id}.json"
        path.unlink(missing_ok=True)
        del self._rubrics[assignment_id]
        log.info("rubrics: deleted %s", assignment_id)

    @property
    def ids(self) -> list[str]:
        return sorted(self._rubrics)

File: frontend/src/lib/rubric.ts
/** Рубрика на стороне редактора.
 *
 *  Здесь три вещи, которых нет в `backend.ts`: повтор серверных проверок,
 *  заготовки пустых сущностей и справочник видов проверок Format Gate.
 *
 *  Про повтор отдельно. Обычно считать на фронте то, что считает сервер, —
 *  ошибка: так появляется правдоподобное, но неверное число рядом с ответом
 *  сервера (ровно поэтому итог работы после правки балла показан
 *  предварительным, а не пересчитан). Здесь случай другой: это не подсчёт, а
 *  список поломок, и он нужен методисту в момент правки, а не после отказа.
 *  Правила скопированы из `backend/src/avito_reviewer/ai/rubric.py`
 *  (`validate_rubric`) вместе с формулировками — чтобы одна и та же поломка не
 *  называлась на экране и в ответе сервера по-разному. **Источник правды —
 *  сервер**: его 422 показывается как есть, даже если здесь проблем не нашлось.
 */

import type { Criterion, FormatCheck, LatePolicy, Rubric } from './backend'
import { plural } from './format'

/** Имя файла рубрики в каталоге — отсюда ограничение на идентификатор.
 *  Экспортируется ради теста, который сверяет его с серверным. */
export const ID_PATTERN = /^[a-zA-Z0-9][a-zA-Z0-9._-]{0,63}$/

const round4 = (value: number): number => Math.round(value * 1e4) / 1e4

/** Число без хвостовых нулей — как `%g` в сообщениях сервера. */
export const g = (value: number): string => String(round4(value))

/** Максимум, который реально можно набрать. Вес 0 сервер читает как 1
 *  (`c.weight or 1.0`), и здесь должно быть так же, иначе экран объявит
 *  недостижимой шкалу, которую сервер примет. */
export function reachableMax(rubric: Rubric): number {
  return round4(
    (rubric.criteria ?? []).reduce((sum, c) => sum + c.max_score * (c.weight || 1), 0),
  )
}

/** Что не так с рубрикой. Пустой список — сервер должен её принять. */
export function rubricProblems(rubric: Rubric): string[] {
  const problems: string[] = []
  const criteria = rubric.criteria ?? []

  if (!ID_PATTERN.test(rubric.assignment_id)) {
    problems.push(
      `идентификатор '${rubric.assignment_id}' не годится для имени файла: ` +
        `латиница, цифры, точка, дефис и подчёркивание`,
    )
  }
  if (!criteria.length) problems.push('в рубрике нет ни одного критерия')

  const ids = criteria.map((criterion) => criterion.id)
  const duplicates = [...new Set(ids.filter((id) => ids.indexOf(id) !== ids.lastIndexOf(id)))].sort()
  if (duplicates.length) {
    problems.push('повторяющиеся идентификаторы критериев: ' + duplicates.join(', '))
  }

  if (rubric.scale.total_max <= 0) problems.push('максимальный балл должен быть больше нуля')
  if (rubric.scale.step < 0) problems.push('шаг шкалы не может быть отрицательным')

  const threshold = rubric.scale.pass_threshold
  if (threshold != null && threshold > rubric.scale.total_max) {
    problems.push(
      `порог зачёта ${g(threshold)} выше максимума ${g(rubric.scale.total_max)} — зачёт недостижим`,
    )
  }

  for (const criterion of criteria) {
    if (criterion.max_score <= 0) {
      problems.push(`${criterion.id}: максимум критерия должен быть больше нуля`)
    }
    const minimum = criterion.min_score_for_pass
    if (minimum != null && minimum > criterion.max_score) {
      problems.push(
        `${criterion.id}: обязательный минимум ${g(minimum)} выше максимума ` +
          `${g(criterion.max_score)} — критерий провален всегда`,
      )
    }
  }

  // Сумма может не совпадать с максимумом: у критериев бывают веса. А вот
  // недостижимый максимум — это уже поломка шкалы.
  const reachable = reachableMax(rubric)
  if (criteria.length && reachable < rubric.scale.total_max) {
    problems.push(
      `максимум ${g(rubric.scale.total_max)} недостижим: по всем критериям ` +
        `с весами набирается ${g(reachable)}`,
    )
  }
  return problems
}

/** Замечания, которые сервер пропускает, а человек потом расхлёбывает.
 *
 *  Держатся отдельно от `rubricProblems` намеренно: те повторяют правила
 *  сервера и потому блокируют отправку, а эти — вопрос вкуса, и решать его
 *  методисту. Смешать их значило бы либо запретить то, что сервер разрешает,
 *  либо утопить настоящую поломку среди придирок. */
export function rubricAdvisories(rubric: Rubric): string[] {
  const notes: string[] = []

  if (!rubric.title.trim()) {
    notes.push('у рубрики нет названия — в каталоге и в выборе рубрики она будет пустой строкой')
  }
  if (!rubric.course.trim()) {
    notes.push('не назван курс — рубрика ляжет в каталог без принадлежности')
  }

  const untitled = (rubric.criteria ?? []).filter((c) => !c.title.trim()).map((c) => c.id)
  if (untitled.length) {
    notes.push(
      `критерии без названия (${untitled.join(', ')}) — в черновике оценки они останутся без заголовка`,
    )
  }
  return notes
}

/** Поля, которых может не быть в ответе, редактору нужны всегда: иначе каждое
 *  обращение к `criteria` или `params` тащит за собой проверку на `undefined`. */
export function normalizeRubric(rubric: Rubric): Rubric {
  return {
    ...rubric,
    stage: rubric.stage ?? null,
    source_note: rubric.source_note ?? '',
    scale: { ...rubric.scale, pass_threshold: rubric.scale.pass_threshold ?? null },
    late_policy: rubric.late_policy ?? { grace_days: 0, penalty_per_grace_day: 0, after_grace: 'zero' },
    format_gate: (rubric.format_gate ?? []).map((check) => ({
      ...check,
      note: check.note ?? '',
      params: { ...(check.params ?? {}) },
    })),
    criteria: (rubric.criteria ?? []).map((criterion) => ({
      ...criterion,
      description: criterion.description ?? '',
      checks: [...(criterion.checks ?? [])],
      anchors: { ...(criterion.anchors ?? {}) },
    })),
  }
}

export function blankRubric(): Rubric {
  return normalizeRubric({
    assignment_id: '',
    title: '',
    course: '',
    stage: null,
    source_note: '',
    ai_policy: 'declare_required',
    scale: { total_max: 10, pass_threshold: null, step: 1 },
    late_policy: { grace_days: 0, penalty_per_grace_day: 0, after_grace: 'zero' },
    format_gate: [],
    criteria: [],
  })
}

/** Свободный `c1`, `c2`, … — идентификаторы критериев обязаны быть уникальны. */
export function nextCriterionId(rubric: Rubric): string {
  const taken = new Set((rubric.criteria ?? []).map((criterion) => criterion.id))
  for (let index = 1; ; index += 1) {
    const candidate = `c${index}`
    if (!taken.has(candidate)) return candidate
  }
}

export function blankCriterion(rubric: Rubric): Criterion {
  return {
    id: nextCriterionId(rubric),
    title: '',
    max_score: 1,
    min_score_for_pass: null,
    weight: 1,
    description: '',
    checks: [],
    anchors: {},
    evidence_required: true,
    auto_verifiable: false,
    ai_sensitive: false,
  }
}

export function blankCheck(): FormatCheck {
  return { check: 'required_paths', level: 'warning', params: { paths: [] }, note: '' }
}

/* ------------------------------------------------------------------------ */

export type ParamKind = 'text' | 'number' | 'list'

export interface ParamField {
  key: string
  label: string
  kind: ParamKind
  hint?: string
}

export interface GateKind {
  check: string
  label: string
  /** Гейт умеет исполнять эту проверку. Остальные не исчезают — уходят
   *  ревьюеру как `inconclusive`, и методист должен знать об этом до того, как
   *  положит проверку в рубрику, а не после первой сдачи. */
  executed: boolean
  hint: string
  params: ParamField[]
}

const PATTERN_FIELD: ParamField = {
  key: 'pattern',
  label: 'Регулярное выражение',
  kind: 'text',
  hint: 'синтаксис Python `re`',
}
const IN_FIELD: ParamField = {
  key: 'in',
  label: 'Только в файлах с расширением',
  kind: 'list',
  hint: 'по одному в строке, например .py',
}

/** Виды проверок, которые понимает `ai/gate.py`. Список закрытый, но не
 *  запретительный: свой `check` завести можно, он просто уйдёт человеку. */
export const GATE_KINDS: GateKind[] = [
  {
    check: 'required_paths',
    label: 'Требуемые пути',
    executed: true,
    hint: 'работа обязана содержать эти файлы или каталоги',
    params: [{ key: 'paths', label: 'Пути', kind: 'list', hint: 'по одному в строке' }],
  },
  {
    check: 'forbidden_paths',
    label: 'Запрещённые пути',
    executed: true,
    hint: 'этих файлов в работе быть не должно — например закоммиченный .env',
    params: [{ key: 'paths', label: 'Пути', kind: 'list', hint: 'по одному в строке' }],
  },
  {
    check: 'code_contains',
    label: 'В работе есть',
    executed: true,
    hint: 'дословное требование условия, которое видно по тексту работы',
    params: [PATTERN_FIELD, { key: 'expected', label: 'То же словами', kind: 'text' }, IN_FIELD],
  },
  {
    check: 'code_absent',
    label: 'В работе не должно быть',
    executed: true,
    hint: 'обратная проверка: совпадение означает провал',
    params: [PATTERN_FIELD, IN_FIELD],
  },
  {
    check: 'token_budget',
    label: 'Объём работы',
    executed: true,
    hint: 'ограничение объёма из условия: страниц в git нет, поэтому считаем токены',
    params: [{ key: 'max_tokens', label: 'Потолок в токенах', kind: 'number' }],
  },
  {
    check: 'revision_history_visible',
    label: 'Видна история изменений',
    executed: false,
    hint: 'требование условия для сдачи через Google Docs; в git-канале не исполняется',
    params: [],
  },
  {
    check: 'font',
    label: 'Шрифт и кегль',
    executed: false,
    hint: 'живёт в .docx и Google Docs, а не в репозитории',
    params: [
      { key: 'family', label: 'Гарнитура', kind: 'text' },
      { key: 'size_pt', label: 'Кегль', kind: 'number' },
    ],
  },
]

export const gateKind = (check: string): GateKind | undefined =>
  GATE_KINDS.find((kind) => kind.check === check)

/* ------------------------------------------------------------------------ */

/** Правило просрочки записано в рубрике числами; человеку нужно предложение.
 *  Живёт здесь, а не на экране рубрик, потому что редактор показывает ту же
 *  фразу предпросмотром: два перевода одного правила разошлись бы. */
export function lateInWords(policy: LatePolicy | undefined): string {
  if (!policy) return 'штраф за просрочку не задан'
  const grace = policy.grace_days ?? 0
  const penalty = policy.penalty_per_grace_day ?? 0
  if (!grace && !penalty) return 'штрафа за просрочку нет'

  const head = grace
    ? `досдача в течение ${grace} ${plural(grace, 'дня', 'дней', 'дней')} — минус ${penalty} ${plural(penalty, 'балл', 'балла', 'баллов')} за день`
    : `минус ${penalty} за каждый день просрочки`
  return head + (policy.after_grace === 'zero' ? ', позже — 0 баллов' : ', дальше штраф растёт')
}

/** У части проверок методист пишет `label`, у части — нет; голое число из
 *  `params` в списке нечитаемо, поэтому у каждого вида проверки есть своя
 *  формулировка. */
export function checkLabel(check: FormatCheck): string {
  const params = (check.params ?? {}) as Record<string, unknown>
  if (typeof params.label === 'string' && params.label) return params.label

  switch (check.check) {
    case 'token_budget':
      return `Объём работы — не больше ${params.max_tokens} токенов`
    case 'required_paths':
      return `Требуемые пути: ${[params.paths].flat().join(', ')}`
    case 'code_absent':
      return `В работе не должно быть: ${params.pattern}`
    default:
      return typeof params.expected === 'string' ? params.expected : check.check
  }
}
